### ISSUE TITLE
Adapter response payloads

### ISSUE THREAD

**Issue by cayuu** [mekanika/adapter#1]
The [Qe specification](https://github.com/mekanika/qe) does not currently provide guidance regarding any expectations of response payloads to any given **Qe** control message. 

On the one extreme, an adapter could simply return `true` or `false` regarding whether it succeeded in applying the Qe. On the other, it could return with detailed Error messages on failure, and return all data changes on success, formatted according to some other specification (eg. JSON-API). Or something in between. 

Standardising this (or providing a mechanism to indicate how its actions respond) would enable machine level interop, automated testing and compatibility, etc. All the good things.

TODO: **Define adapter responses**

**Comment by cayuu** [comment]
**[JSON-API](http://jsonapi.org)** is a good starting point.

In particular when considering `.populate` _Qe_, the main question is how to return results (either embedded in primary results or separately as their own resultsets). eg. (pseduo-results, not formatted according to any standard):

``` js
// Embedded
[
  { id:1, title:"Long title", body:"whatever", comments: [
    {id:50, msg:"comment1"}, {id:51, msg:"comment2"}
  ]}
]
// vs. something like:
{
  posts: [
    { id:1, title:"Long title", body:"whatever", comments: [50,51]}
  ],
  comments: [
      {id:50, msg:"comment1"}, {id:51, msg:"comment2"}
  ]
}
```

The latter approach is closer to JSON-API and benefits from being able to push merge complexity into clients (ie. a plugin to 'denormalise' into 'embedded' form).

**Comment by cayuu** [comment]
Not specifically Adapter related, but some W3C recommendations for **hypermedia** related matters:
- [JSON-LD **JSON-based serialisation for Linked Data**](http://www.w3.org/TR/json-ld/): Of particular interest are the `@id`, `@context` and `@graph` keywords used to describe the underlying models and relations
- [RDF **Resource Description Framework**](http://www.w3.org/TR/2014/PR-rdf11-concepts-20140109/): Graph based framework defining data models. (This is way overkill for what's needed here).

**Comment by cayuu** [comment]
What is required is a convention for Adapters to adhere to. Allowing other plugins in the response stack/pipeline to convert data formats makes a lot of sense.

Some formatting questions
- Representing responses as array vs. object (when/why)
- Keyed response objects (find `posts` - return `{posts: [...]}` vs return `[...]`)
- Relation/linked data representation (embedded vs side-load)

For **related/linked data** the question more specifically is: once instructed to side-load related data (via a _Qe_ `.populate` field), should that related data be embedded within the primary records or provided separately.  

**tl;dr**: Separately representing related/linked data seems the best approach. Further details in next posts

---

> **A sidebar on Schema**: Given an application level schema for a Resource, any Store-level implementation that differs from the application level schema will need to supply and/or maintain a map back to application level expectations. (e.g. Store-level implementations of the schema may be implemented quite differently for performance reasons per-store).

**Comment by cayuu** [comment]
The possibility of side-loaded data prefers returning all response payloads as `Objects` (not `Arrays`), which requires the primary response to be keyed.

``` js
{
  primary_response: $responsePayload
}
```

The key point to note is that this request will _always_ be in response to a _Qe_. Which means any side-loading directive **MUST** come in the form of a `.populate` directive. This specifies `fields` to populate on. 

Assuming a known `.on` _Qe_ field, and a given populate directive:

``` js
var Qe = {do:'find', on:'posts', populate:{tags:{}}}
// Returns:
{
  posts: [
    {id:1, title:"Hello world", tags:[512, 224]}
  ],
  tags: [
    {id:512, label:"special"},
    {id:224, label:"new"}
  ]
}
```

There exists potential for conflict here _if_ the resource has a field of the same name as the resource (ie. `posts` has a `posts` field eg. `posts.posts`) or if deeply nested relations have the same keys. The only way to ensure non-collision is to namespace the linked keys, either by turning them into a string `"posts.posts": [...]` or by putting them into a separate collection `linked: { posts: [...] }`. The latter introduces a reserved/convention word (in this case `linked`).

``` js
var Qe = {do:'find', on:'posts', populate:{tags:{}}}
// Returns:
{
  posts: [
    {id:1, title:"Hello world", tags:[512, 224]}
  ],
  linked: {
    tags: [
      {id:512, label:"special"},
      {id:224, label:"new"}
    ]
  }
}
```

Where `.on` is NOT specified, defaulting to `data` as the primary response key seems appropriate.

It is the responsibility of some other piece in the chain to compose the Adapter response (whatever that may be) into any given "final" format.

Several questions remain: 
- how to represent singular response objects (ie. "one" post): `{do:'find', on:'posts', ids:[1]}`. Or rather, assuming the current preference for "always array", better to ask: "under what conditions would it be desirable to respond with an object rather than an array?"
- how to normalise a Store that responds `true|false` to a request (often applicable to actions like `update` and `remove`)?
- how to (if at all) represent responses for non-resource requests (like `count` or `sum`) Note that these types of requests are not explicitly supported at the Qe level. (see ticket at [Qe#7](https://github.com/mekanika/qe/issues/7))
- how to present deeply nested data (`posts.comments.author.friends.likes`), flattened or nested...

**Comment by cayuu** [comment]
### Success Responses

On a successful application of the _Qe_ action, the options for responses are generally:
1. `Boolean` true
2. `Number` of affected/matched
3. `Array` of identifiers affected/matched
4. `Array` of records 

Various stores may support differing levels of these response per method (CRUD related). As such, mandating that all methods always return option 4 (array of complete elements) may _force_ adapters to make multiple calls into a Store that does not natively support that option.

Initially, adapters could expose `supported` set of flags defining which option is returned per operation. Thus allowing the application to decide what/how to manage. Eg:

``` js
var BOOL = 1, NUMBER = 2, IDS = 3, COMPLETE = 4;
var defaultResponses = {
  create: COMPLETE,
  update: NUMBER,
  find: COMPLETE,
  remove: NUMBER,
}
```

Which raises a consideration - say an `update` `.match` matches 50,000 records - we probably _don't_ want 50k `COMPLETE` records dumped back up the stack. MongoDB deals with this by simply returning the number of records that were affected.

It's possible to provide for overriding the default limit:

``` js
// Give me the million records I'm asking for
// (hope you have the RAM for this)
{ meta: {ignoreLimit:true} }
```

(plz can we streams much?)

If a store provides no mechanism for identifying modified records, returning instead ONLY the number of affected records/rows/documents, then tracking query-matched records in a worst-case requires that a `find` operation is done _first_ (get the record ids) and _then_ an `update` (returns the number of updated), and _then_ a `find` based on the ids of the initial find to get the now updated records. It may be necessary to lock the records to ensure atomicity in high write-load environments.

If a downstream application component knows an Adapter's response capabilities, it could potentially provide fine-grain _Qe_ control via the `.meta` flag or modify its use of the Adapter accordingly.

**Comment by cayuu** [comment]
For reference - I'm leaning away from modifying adapter behaviour on the fly via _Qe_. Adapter setups should be configured at the application level, with direct control.
